### PR TITLE
ARROW-15614: [C++] Add sqrt binary scalar kernel

### PR DESCRIPTION
**Pull request by rtpsw**
This is a proposal for resolving [ARROW-15614](https://issues.apache.org/jira/browse/ARROW-15614). Tests will be added when the resolution approach is agreed

### CODE REVIEW

**Review by lidavidm — COMMENTED** [review]
This looks good to me.

For tests, we just need to add cases in scalar_arithmetic_test.cc as part of the `TestUnaryArithmeticFloating`, `TestUnaryArithmeticIntegral`, and `TestUnaryArithmeticDecimal` test cases:
https://github.com/apache/arrow/blob/3cc7df895e0cf2bf0aa1ac1977c20e5f93ec2f25/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc#L3012

(since integral/decimal will get cast to float first)

**Review by lidavidm — APPROVED** [review]
LGTM, thank you! Let's see how CI goes

**Comment by lidavidm** [comment]
Looks like there's a few lint issues, can you either apply the changes from the pipeline (https://github.com/apache/arrow/runs/5248147063?check_suite_focus=true#step:5:787) or run the formatter (https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci)?

**Comment by lidavidm** [comment]
Argh. IIRC, MinGW's floating point implementations are…less precise.

**Comment by ianmcook** [comment]
Thanks for doing this! @rtpsw could you please add an entry for this in the table of **Arithmetic functions** in [compute.rst](https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst).

**Comment by rtpsw** [comment]
> Thanks for doing this! @rtpsw could you please add entries for `sqrt` and `sqrt_checked` in the table of **Arithmetic functions** in [compute.rst](https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst)?

Will do.

@lidavidm: As for the tests, now they succeed on "AMD64 Windows MinGW 64" but fail on "AMD64 Windows MinGW 32". This is weird, given that I added a tolerance. How would you suggest resolving this? It's not going to be convenient for me to set up the kind of system on which the tests fail.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc [hunk 1]
```diff
@@ -3284,6 +3312,35 @@ TYPED_TEST(TestUnaryArithmeticSigned, Log) {
...
+  auto max_val = std::numeric_limits<CType>::max();
+  for (auto check_overflow : {false, true}) {
+    this->SetOverflowCheck(check_overflow);
+    this->AssertUnaryOp(Sqrt, "[1, 2, null, NaN, Inf]",
+                        "[1, 1.414213562, null, NaN, Inf]");
+    this->AssertUnaryOp(Sqrt, min_val, static_cast<CType>(std::sqrt(min_val)));
+    double atol = this->equal_options_.atol();
+    this->equal_options_.atol(max_val / 1e12);
+    this->AssertUnaryOp(Sqrt, max_val, static_cast<CType>(std::sqrt(max_val)));
+    this->equal_options_.atol(atol);
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:3338]: It's preserved for the second loop iteration.

**Comment by rtpsw** [comment]
> Also, sorry, it should go in python/api/compute.rst as well: https://github.com/apache/arrow/blob/master/docs/source/python/api/compute.rst#arithmetic-functions

Fixing this shortly.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc]: Ah, whoops, indeed.

**Review by lidavidm — APPROVED** [review]

**Comment by rtpsw** [comment]
This time the "AMD64 Windows MinGW 64" failures are unrelated to the current code changes.

**Comment by ianmcook** [comment]
> This time the "AMD64 Windows MinGW 64" failures are unrelated to the current code changes.

I re-ran the tests and everything is passing now

**Comment by rtpsw** [comment]
Is the PR good to go?

**Comment by lidavidm** [comment]
Thanks for the reminder, and thank you for implementing this!

**Pull request closed without merging**